### PR TITLE
Revise calc zone sums

### LINKED ISSUE #8203: Speedup CalcZoneSums
Issue overview
--------------
LBL identified 20 hotspots for speed improvement. CalcZoneSums is one of hotspots. The function is called by Predictor and Corrector every time step. A section of supply system calculation is not needed.   

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by lgu1234**
Pull request overview
---------------------
A new default argument is added. The CalcZoneSums function is called by predictor and corrector.  When the predictor calls the function with the new argument, the section of supply system calculation is bypassed to speed up performance, because the supply system calculation is not necessary.  

Resolves #8203

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] ~If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix~
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

### CODE REVIEW

File: src/EnergyPlus/ZoneTempPredictorCorrector.hh [hunk 1]
```diff
@@ -284,7 +284,8 @@ namespace ZoneTempPredictorCorrector {
                       Real64 &SumMCp,     // Zone sum of MassFlowRate*Cp
                       Real64 &SumMCpT,    // Zone sum of MassFlowRate*Cp*T
                       Real64 &SumSysMCp,  // Zone sum of air system MassFlowRate*Cp
-                      Real64 &SumSysMCpT  // Zone sum of air system MassFlowRate*Cp*T
+                      Real64 &SumSysMCpT,             // Zone sum of air system MassFlowRate*Cp*T
+                      bool const CorrectorFlag = true // Corrector call flag
```
Comment (Myoldmopar) [inline src/EnergyPlus/ZoneTempPredictorCorrector.hh:286]: Using a default value is fine, but it's also OK to just require passing it in, that way any client code has to intentionally set the arg to true or false.  No big deal.

File: src/EnergyPlus/ZoneTempPredictorCorrector.cc [hunk 2]
```diff
@@ -6203,84 +6204,81 @@ namespace ZoneTempPredictorCorrector {
         // Sum all system air flow: SumSysMCp, SumSysMCpT
         // Check to see if this is a controlled zone
         ControlledZoneAirFlag = Zone(ZoneNum).IsControlled;
+        if (CorrectorFlag) {
```
Comment (Myoldmopar) [inline src/EnergyPlus/ZoneTempPredictorCorrector.cc:6176]: So if we are just in predictor, we skip a bunch of stuff.  I'm fine with that conceptually, and if CI is showing no changes in results then that's enough confirmation.

File: tst/EnergyPlus/unit/ZoneTempPredictorCorrector.unit.cc [hunk 3]
```diff
@@ -1059,6 +1058,16 @@ TEST_F(EnergyPlusFixture, ZoneTempPredictorCorrector_CalcZoneSums_SurfConvection
...
+
+    Node(1).MassFlowRate = 0.1;
+    Node(2).MassFlowRate = 0.2;
+    CalcZoneSums(state.dataZonePlenum, ZoneNum, SumIntGain, SumHA, SumHATsurf, SumHATref, SumMCp, SumMCpT, SumSysMCp, SumSysMCpT);
+    EXPECT_NEAR(302.00968500, SumSysMCp, 0.0001);
+    EXPECT_NEAR(6040.1937, SumSysMCpT,0.0001);
+
+    CalcZoneSums(state.dataZonePlenum, ZoneNum, SumIntGain, SumHA, SumHATsurf, SumHATref, SumMCp, SumMCpT, SumSysMCp, SumSysMCpT, false);
+    EXPECT_EQ(0.0, SumSysMCp);
+    EXPECT_EQ(0.0, SumSysMCpT);
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ZoneTempPredictorCorrector.unit.cc:1070]: Seems like a sufficient unit test.  In the first case we are doing the corrector and we calculate the system mCp and mCpT terms.  In the second case we aren't doing the corrector and so the values come back zero for the system.

**Review by Myoldmopar — APPROVED** [review]
This seems like a decent set of changes from purely a code review standpoint.

**Comment by Myoldmopar** [comment]
CI results of interest:
- ASHRAE9012016_Hospital_Denver has ERR, ESO Small, and MTR Small diffs.  From the err diff it looks like things are just slightly off, at a small fraction of a degree Celsius, in some warning messages.
- ASHRAE9012016_OfficeLarge_Denver has EIO and ERR diffs that are again, tiny differences in warmup convergence and error messages
- FourPipeBeamLargeOffice has ERR diffs that are again, just small changes in error messages
- Hundreds and hundreds of files with no diffs at all, which is extremely promising for this change
- Performance moved in the right direction
- Unit tests all pass, including new one

I feel like this is probably OK to drop in, but I'll do a local build to make sure it doesn't need updating from recent develop.  Anyone have any additional thoughts/concerns before I move to that step?

**Comment by mjwitte** [comment]
@Myoldmopar Sooo, if the sums that are now skipped truly aren't used anywhere except in corrector, then why are there any diffs at all? Is there some component that using these sums directly perhaps?

**Comment by Myoldmopar** [comment]
> Is there some component that using these sums directly perhaps?

That would be my guess, and the error diffs are all with the reformulated EIR chiller.  Suspicious...yes?

**Comment by Myoldmopar** [comment]
@lgu1234 can you report on why these changes might cause diffs in the chiller?  Maybe just inspecting the reformulated chiller code and seeing if it directly accesses some underlying values.

**Comment by lgu1234** [comment]
@Myoldmopar I am going to check.

**Comment by lgu1234** [comment]
@Myoldmopar @mjwitte I found the cause. The difference is caused by my consolidation of CpAir calculation. The consolidate one is  CpAir = PsyCpAirFnW(ZoneAirHumRat(ZoneNum));

There are two types of CpAir calculation in the original version:

1.                 CpAir = PsyCpAirFnW(ZoneAirHumRat(ZoneNum));
2.             Real64 const air_hum_rat(ZoneAirHumRat(ZoneNum));
                CpAir = PsyCpAirFnW(air_hum_rat);

Therefore, the real difference is that I replace "Real64 const" by "Real64". After merge this morning by reverting my consolidation, CI tests are all green.

**Comment by mjwitte** [comment]
@lgu1234 That is strange. I believe you, but I don't understand how that makes any difference.

**Comment by Myoldmopar** [comment]
Github is showing this in a funny order, but 5fdbd28 is the last commit on here, and it was all green now.  I think this is OK now, but we can pull develop in one more time to be sure.

**Comment by Myoldmopar** [comment]
Built and ran with develop pulled in.  Unit tests all pass fine.  I pushed and we'll let CI run one last time before doing anything else.

**Comment by nrel-bot-3** [comment]
@lgu1234 @lgentile it has been 28 days since this pull request was last updated.

**Comment by mjwitte** [comment]
@Myoldmopar Assuming windows CI comes back clean, this is ready to drop in. Any further comments?

**Comment by Myoldmopar** [comment]
Yeah this is great.  Thanks @lgu1234 and @mjwitte.  Merging.

**Pull request merged**